_posts/2019-09-19-haproxy-mala-rzecz-a-cieszy.md

```diff
@@ -0,0 +1,229 @@
+Poza klasycznym użyciem jako reverse-proxy zapewniające load-balancing i high-availability, HAProxy może być z powodzeniem wykorzystane w warsztacie programistycznym. 
+Przedstawię poniżej kilka zastosowań z własnego doświadczenia.
+
+### Routing, który łatwo zmienić
+Rozwijając aplikację, która wywołuje serwisy z innych aplikacji, chcemy niekiedy móc przełączać się między różnymi adresami tych zewnętrznych usług, 
+np. możemy chcieć sprawnie przełączać się między prawdziwymi aplikacjami oraz ich zmockowanymi wersjami. 
+Często restart naszej aplikacji jest długotrwały, a zmiana namiarów na zewnętrzne usługi wymaga edycji więcej niż jenego pliku.
+W takim przypadku HAProxy może nam posłużyć jako wygodna "centralka", w której będziemy się sprawnie przełączać między różnymi wersjami zewnętrznych usług.
+Konfigurując HAProxy w ten sposób:
+```
+frontend my-forward-proxy
+    bind 172.19.0.1:81
+    default_backend my-external-services
+    
+backend my-external-services
+    server my-external-server-1 172.19.0.2:8080
+
+backend my-mocked-services
+    server my-mocked-server-1 172.19.0.3:8080
+```
+i ustawiając w naszej aplikacji adres zewnętrznych usług na 172.19.0.1:81, 
+możemy bez restartu aplikacji zmieniać jej zewnętrzne zależności przez zmianę wartości "default_backend" i szybki "reload" HAProxy.
+
+### Podglądanie SSL'a
+Jeśli serwisy w naszym systemie przesyłają sobie dane przez szyfrowane połączenia TLS/SSL, może to stanowić przeszkodę w debugowaniu komunikacji przy pomocy analizatorów ruchu sieciowego, 
+takich jak tcpdump, czy Wireshark. 
+Możemy jednak pomiędzy serwisami ustawić pośrednika w postaci HAProxy, które będzie terminowało szyfrowane połączenie i umożliwiało nam podglądanie ruchu sieciowego.
+
+### Opóźnienia requestów
+Podczas rozwijania aplikacji zachodzi czasem potrzeba przetestowania jak zachowa się ona w przypadku dłuższego niż oczekiwane przetwarzania w wywoływanych usługach zewnętrznych.
+Jeśli pośrednikiem w dostępie do zewnętrznych usług jest HAProxy, to mamy możliwość zasymulowania takiego opóźnienia.
+HAProxy samo w sobie nie posiada takiej funkcjonalności, ale pozwala na rozszerzanie możliwości za pomocą skryptów pisanych w języku Lua.
+Przykładowy skrypt dodający opóźnienie:
+```
+function delay_request(txn)
+    core.msleep(15000)
+end
+
+core.register_action("delay_request", { "http-req" }, delay_request);
+```
+Załadowanie skryptu do HAProxy:
+```
+    lua-load /etc/haproxy/delay.lua
+```
+i użycie w konfiguracji:
+```
+frontend my-delay-frontend
+    bind 172.19.0.1:81
+    mode http
+    http-request lua.delay_request
+    default_backend my-delay-backend
+
+backend my-delay-backend
+    server my-external-server-1 172.19.0.2:8080
+```
+### Prosty serwer statycznej zawartości
+W dość przewrotny sposób możemy wykorzystać HAProxy jako prosty serwer statycznych plików. 
+Dla każdego backendu możemy skonfigurować pliki, które mają zostać zaserwowane w przypadku określonych statusów HTTP.
```

Comment:
W artykule wszędzie jest backend'u a tutaj bez apostrofa. Należałoby uspójnić.

Ja prywatanie jestem fanem "spolszania" pisowni tam gdzie się da i odniemiałbym pisząć "backendu" i "frontendu". Czy dalej "toolboxa". Apostrof stosowałbym w ostateczności.

Reply:
Tutaj jest to fajnie rozpisane: http://wittamina.pl/apostrof-w-zlym-miejscu/